Snowflake.Data/Core/CredentialManager/Infrastructure/SFCredentialManagerFileImpl.cs

```diff
@@ -0,0 +1,147 @@
+        {
+            var customDirectory = _environmentOperations.GetEnvironmentVariable(CredentialCacheDirectoryEnvironmentName);
+            _jsonCacheDirectory = string.IsNullOrEmpty(customDirectory) ? Path.Combine(HomeDirectoryProvider.HomeDirectory(_environmentOperations), CredentialCacheDirName) : customDirectory;
+            if (!_directoryOperations.Exists(_jsonCacheDirectory))
+            {
+                _directoryOperations.CreateDirectory(_jsonCacheDirectory);
+            }
+            _jsonCacheFilePath = Path.Combine(_jsonCacheDirectory, CredentialCacheFileName);
+            s_logger.Info($"Setting the json credential cache path to {_jsonCacheFilePath}");
+        }
+
+        internal void WriteToJsonFile(string content)
+        {
+            s_logger.Debug($"Writing credentials to json file in {_jsonCacheFilePath}");
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                _fileOperations.Write(_jsonCacheFilePath, content);
+            }
+            else
+            {
+                if (!_directoryOperations.Exists(_jsonCacheDirectory))
+                {
+                    _directoryOperations.CreateDirectory(_jsonCacheDirectory);
+                }
+                s_logger.Info($"Creating the json file for credential cache in {_jsonCacheFilePath}");
+                if (_fileOperations.Exists(_jsonCacheFilePath))
+                {
+                    s_logger.Info($"The existing json file for credential cache in {_jsonCacheFilePath} will be overwritten");
+                }
+                var createFileResult = _unixOperations.CreateFileWithPermissions(_jsonCacheFilePath,
+                    FilePermissions.S_IRUSR | FilePermissions.S_IWUSR);
+                if (createFileResult == -1)
+                {
+                    var errorMessage = "Failed to create the JSON token cache file";
+                    s_logger.Error(errorMessage);
+                    throw new Exception(errorMessage);
+                }
+                else
+                {
+                    _fileOperations.Write(_jsonCacheFilePath, content);
+                }
+
+                var jsonPermissions = _unixOperations.GetFilePermissions(_jsonCacheFilePath);
+                if (jsonPermissions != (FileAccessPermissions.UserRead | FileAccessPermissions.UserWrite))
+                {
+                    var errorMessage = "Permission for the JSON token cache file should contain only the owner access";
+                    s_logger.Error(errorMessage);
+                    throw new Exception(errorMessage);
+                }
+            }
+        }
+
+        internal KeyTokenDict ReadJsonFile()
+        {
+            var contentFile = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? File.ReadAllText(_jsonCacheFilePath) : _fileOperations.ReadAllText(_jsonCacheFilePath, TomlConnectionBuilder.ValidateFilePermissions);
```

Comment:
don't use TomlConnectionBuilder.ValidateFilePermissions because token cache has nothing to do with toml. It is better to create similar function here because we would not like to accidentally change TokenCache file validations if requirements for Toml change.
If the method would be exaclty the same and you would like to avoid code duplication you could create a common method for that eg. ValidateFileIsAccessedOnlyByItsOwner and you could call this method from both of validation methods.

Reply:
Moved method to UnixOperations class